File: 프로그래머스/lv3/43162. 네트워크/네트워크.py
# 아이디어 : 1. visited 테이블 생성
#           2. 해당 노드를 한 번도 방문한 적이 없다면 dfs 탐색
#           3. dfs에서 다른 노드를 방문한 적이 없으면서 연결되어 있다면
#               재귀적으로 dfs 실행
# 알고리즘 : dfs
# 자료구조 : stack(recursion)

def solution(n, computers):
    answer = 0
    # visited 테이블 생성
    visited = [False for _ in range(n)]
    
    # dfs 함수 정의
    def dfs(graph, v, visited):
        # 현재 노드는 방문 처리
        visited[v] = True
        
        for i in range(n):
            # 해당 노드를 방문한 적이 없으면서 연결되어 있다면
            if not visited[i] and graph[v][i] == 1:
                dfs(graph, i, visited)
                
    for v in range(n):
        # 해당 노드를 방문한 적이 없다면 dfs 실행과 네트워크 추가
        if not visited[v]:
            dfs(computers, v, visited)
            answer += 1
    
    return answer

File: 프로그래머스/3/43162. 네트워크/네트워크.java
// 아이디어: visited 테이블 + 너비 우선탐색 하기
// 알고리즘: bfs
// 자료구조: queue

import java.util.*;

class Solution {
    public int solution(int n, int[][] computers) {
        int answer = 0;
        boolean[] visited = new boolean[n];
        
        for (int i = 0; i < n; i++) {
            visited[i] = false;
        }
        
        for (int i = 0; i < n; i++) {
            // 한 번도 방문한 적이 없다면
            if (!visited[i]) {
                bfs(n, i, computers, visited);
                answer += 1;
            }
        }
        
        return answer;
    }
    
    public void bfs(int n, int start, int[][] computers, boolean[] visited) {
        // 큐 및 시작 노드 정의
        Queue<Integer> queue = new LinkedList<>();
        queue.offer(start);
        visited[start] = true;
        
        while (!queue.isEmpty()) {
            int now = queue.poll();

            for (int i = 0; i < n ; i++) {
                // 한 번도 방문한 적이 없으면서 연결되어 있다면
                if (!visited[i] && computers[now][i] == 1) {
                    queue.offer(i);
                    visited[i] = true;
                }
            }
        }
        
    }
}